### PR TITLE
Fixes #13516 - fix ValueError and AssertionError to provide better debugging information

### LINKED ISSUE #13516: For Pulumi Python SDK add to ValueError and AssertionError messages information about resource_name and argumentname
### What happened?

Debugging in Python SDK which resource is failing could be a nightmare because Traceback is leading to pulumi-language-python-exec and not to user program code. Error messages like 'ValueError('unexpected input of type Dummy')' are not very useful.
### 

### Expected Behavior

I would like to get an error message containing information about the resource causing the issue and witch it's property is causing an issue in the format like:

ValueError: During processing resource: 'pda_ec1_devops_test_1_flowlog_BucketOwnershipControls_b' of type <pulumi_aws.s3.bucket_ownership_controls.BucketOwnershipControls object at 0x7fd5323d6cd0> ValueError has risen: 'ValueError("During processing input field called: 'bucket' of type <config_to_resources.Dummy object at 0x7fd5327ca5b0> ValueError has risen: 'ValueError('unexpected input of type Dummy')'")'

### Steps to reproduce

try to create resources like this:
```
class Dummy:
    def __init__(self):
        pass


dd = Dummy()

pulumi_aws.s3.BucketOwnershipControls(
        resource_name="pda_ec1_devops_test_1_flowlog_BucketOwnershipControls_c",
        args=pulumi_aws.s3.BucketOwnershipControlsArgs(
            bucket=dd,
            rule=pulumi_aws.s3.BucketOwnershipControlsRuleArgs(
                object_ownership="BucketOwnerEnforced"
            )
        )
    )
// or this
pulumi_aws.s3.Bucket(
    resource_name="pda_ec1_devops_test_1_flowlog_s3",
    args=pulumi_aws.s3.BucketArgs(
        bucket="some-bucket",
        lifecycle_rules=[
            pulumi_aws.s3.BucketLifecycleRuleArgs(
                enabled=True,
                expiration=pulumi_aws.s3.BucketLifecycleRuleExpirationArgs(
                    days=365
                ),
                id="CleanupRule"
            )
        ],
        tags=[{ #issue here list instead of dict when you mix pulumi_aws with pulumi_aws_native 
            "Name": "some-bucket",
            "SomeTag": "xyz"
        }],
    )
)
```

### Output of `pulumi about`

CLI          
Version      3.74.0
Go Version   go1.20.5
Go Compiler  gc

Plugins
NAME        VERSION
aws         5.41.0
aws-native  0.67.0
python      unknown

Host     
OS       ubuntu
Version  20.04
Arch     x86_64

This project is written in python: executable='/home/adgo/.pyenv/shims/python3' version='3.8.10
'

Current Stack: name

Found no resources associated with name

Found no pending operations associated with name

Backend        
Name         ***
URL            s3
User           ***
Organizations  

Dependencies:
NAME                      VERSION
boto3                     1.26.165
mypy                      1.4.1
pip                       23.1.2
pkg_resources             0.0.0
pulumi-aws                5.41.0
pulumi-aws-native         0.67.0
pydantic                  2.0.2
service-directory-loader  1.0.4
setuptools                44.0.0

### Additional context

Proposed improvements
1)
https://github.com/pulumi/pulumi/blob/86a240362955ecf64752841d65853194ad8a9a44/sdk/python/lib/pulumi/runtime/rpc.py#L209
wrap with try ... except block like : 
```
        try:
            result = await serialize_property(
                v, deps, input_transformer, get_type(k), keep_output_values
            )
        except ValueError as e: 
            raise ValueError(f"During processing input field called: {repr(k)} of type {repr(v)} ValueError has risen: '{repr(e)}'").with_traceback(e.__traceback__)
        except AssertionErroras e: 
            raise AssertionError(f"During processing input field called: {repr(k)} of type {repr(v)} ValueError has risen: '{repr(e)}'").with_traceback(e.__traceback__)
```
2)
https://github.com/pulumi/pulumi/blob/86a240362955ecf64752841d65853194ad8a9a44/sdk/python/lib/pulumi/runtime/resource.py#L872
wrap with try ... except block like : 
```
            try:
                resolver = await prepare_resource(res, ty, custom, remote, props, opts, typ)
            except ValueError  as e: 
                raise ValueError(f"During processing resource: {repr(name)} of type {repr(res)} ValueError has risen: '{repr(e)}'").with_traceback(e.__traceback__)
            except AssertionError: as e:
                raise AssertionError:(f"During processing resource: {repr(name)} of type {repr(res)} ValueError has risen: '{repr(e)}'").with_traceback(e.__traceback__)
```
### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already). 

see additional content with proposed improvements

### PR DESCRIPTION
**Pull request by gothrek22**
fix ValueError and AssertionError to provide better debugging information

### CODE REVIEW

**Review by brandonpollack23 — APPROVED** [review]

**Comment by gothrek22** [comment]
@brandonpollack23 hi, could you approve a rerun? I've fixed the two linting issues in sdk.

**Comment by gothrek22** [comment]
My bad, I must've missed that the linter "auto-fixed" some of the issues that made the tests fail. They should pass this time, I've double checked locally.

@brandonpollack23 thanks for your help mate.

**Comment by gothrek22** [comment]
@brandonpollack23 they seem to have passed successfully.

File: changelog/pending/20241018--sdk-python--fixes-13516-fix-valueerror-and-assertionerror-to-provide-better-debugging-information.yaml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+changes:
+- type: fix
+  scope: sdk/python
+  description: "Fixes #13516 - fix ValueError and AssertionError to provide better debugging information"
```
Comment (tgummerer) [inline changelog/pending/20241018--sdk-python--fixes-13516-fix-valueerror-and-assertionerror-to-provide-better-debugging-information.yaml:4]:
```suggestion
  description: "Add additional debug information to `ValueError` and `AssertionError`"
```

**Review by tgummerer — APPROVED** [review]

**Pull request merged**